gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/ClassUniquenessAnalyzer.java

```diff
@@ -0,0 +1,113 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.tasks;
+
+import java.io.File;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.slf4j.Logger;
+
+public final class ClassUniquenessAnalyzer {
+
+    private final Map<String, Set<ModuleVersionIdentifier>> classToJarsMap = new HashMap<>();
+    private final Map<Set<ModuleVersionIdentifier>, Set<String>> jarsToClasses = new HashMap<>();
+    private final Logger log;
+
+    public ClassUniquenessAnalyzer(Logger log) {
+        this.log = log;
+    }
+
+    public void analyzeConfiguration(Configuration configuration) {
+        Instant before = Instant.now();
+
+        Set<ResolvedArtifact> dependencies = configuration
+                .getResolvedConfiguration()
+                .getResolvedArtifacts();
+
+        Map<String, Set<ModuleVersionIdentifier>> tempClassToJarMap = new HashMap<>();
+        dependencies.stream().forEach(resolvedArtifact -> {
+            File file = resolvedArtifact.getFile();
+            if (!file.exists()) {
+                log.info("Skipping non-existent jar {}: {}", resolvedArtifact, file);
+                return;
+            }
+
+            try (JarFile jarFile = new JarFile(file)) {
+                Enumeration<JarEntry> entries = jarFile.entries();
+                while (entries.hasMoreElements()) {
+                    JarEntry jarEntry = entries.nextElement();
+
+                    if (!jarEntry.getName().endsWith(".class")) {
+                        continue;
+                    }
+
+                    multiMapPut(tempClassToJarMap,
+                            jarEntry.getName().replaceAll("/", ".").replaceAll(".class", ""),
+                            resolvedArtifact.getModuleVersion().getId());
+                }
+            } catch (Exception e) {
+                log.error("Failed to read JarFile {}", resolvedArtifact, e);
+                throw new RuntimeException(e);
+            }
+        });
+
+        tempClassToJarMap.entrySet().stream()
+                .filter(entry -> entry.getValue().size() > 1)
+                .forEach(entry -> {
+                    // add to the top level map
+                    entry.getValue().forEach(value -> multiMapPut(classToJarsMap, entry.getKey(), value));
```

Comment:
not sure why we need to save off this collection if we're not returning it anywhere -- getProblemJars could be implemented with jarsToClasses.keySet(), i believe